### PR TITLE
SISRP-30513 - Grading My Academics merged model reporting errors in QA

### PR DESCRIPTION
**Pull request by davidlimcheng**
https://jira.berkeley.edu/browse/SISRP-30513

* Really glad this was caught, but a bit worried that it was caught during load testing and not beforehand.
* Fixed an error check that was only checking for Final grade status.
* Needed to fortify the code a bit more for nil values from APIs, and fudged on an error object that needed to include both final and midpoint values.
* QA PR to come upon merge.

### CODE REVIEW

File: app/models/my_academics/grading.rb [hunk 1]
```diff
@@ -250,12 +250,16 @@ def parse_cc_grading_status(cs_grading_status, is_law, is_midpoint, term_code)
...
 
     def unexpected_cs_status?(cs_grading_status, is_law)
-      return false if !!%w{GRD RDY APPR POST}.find { |s| s == cs_grading_status[:finalStatus] } || cs_grading_status[:finalStatus].blank?
-      if !is_law
-        return false if !!%w{APPR NRVW RDY}.find { |s| s == cs_grading_status[:midpointStatus] } || cs_grading_status[:midpointStatus].blank?
+      has_error = false
+      if cs_grading_status.nil? ||
+         !(!!%w{GRD RDY APPR POST}.find { |s| s == cs_grading_status.try(:[],:finalStatus) } || cs_grading_status.try(:[],:finalStatus).blank?) ||
+         (!is_law && !(!!%w{APPR NRVW RDY}.find { |s| s == cs_grading_status.try(:[],:midpointStatus) } || cs_grading_status.try(:[],:midpointStatus).blank?))
+        has_error = true
```
Comment (lyttam) [inline app/models/my_academics/grading.rb]: @davidlimcheng since it's late in the game, this is more a future suggestion than anything - but it would be nice to have this logic encapsulated in functions with useful names that describe the meaning of the logic.  For example, 

```
def unexpected_cs_status?(cs_grading_status, is_law)
    has_error = false
    if cs_grading_status.nil? || !special?(cs_grading_status) || . . . 
end

def special?(cs_grading_status)
    (!!%w{GRD RDY APPR POST}.find { |s| s == cs_grading_status.try(:[],:finalStatus) }
end
```

Replacing 'special' with some word that describes what `[GRD, RDY, APPR, POST]` means in plain English.

Comment (davidlimcheng) [inline app/models/my_academics/grading.rb]: @lyttam That's a great suggestion and makes this much, much more readable.  I'd actually like to change this now, it shouldn't take too long.  I'll amend.

File: app/models/my_academics/grading.rb [hunk 2]
```diff
@@ -250,12 +250,16 @@ def parse_cc_grading_status(cs_grading_status, is_law, is_midpoint, term_code)
...
       end
-      logger.debug "Unexpected CS Final Grading Status Received (Final: #{cs_grading_status[:finalStatus]}#{', Midpoint: ' + cs_grading_status[:midpointStatus] if cs_grading_status.key?(:midpointStatus)}) for Class #{self.class.name} feed, uid = #{@uid}"
-      true
+      if has_error && cs_grading_status.present?
+        logger.debug "Unexpected CS Final Grading Status Received (Final: #{cs_grading_status[:finalStatus]}#{', Midpoint: ' + cs_grading_status[:midpointStatus] if cs_grading_status.key?(:midpointStatus)}) for Class #{self.class.name} feed, uid = #{@uid}"
```
Comment (redconfetti) [inline app/models/my_academics/grading.rb]: Did you test this logger statement in some manner?

Comment (redconfetti) [inline app/models/my_academics/grading.rb]: @davidlimcheng - Actually this works, I just have never used a conditional statement within an interpolated statement like this, I wasn't sure how it would work - https://repl.it/Fm8A
